test/configCases/plugins/virtual-url-plugin/webpack.config.js

```diff
@@ -0,0 +1,40 @@
+const fs = require("fs");
+const path = require("path");
+const webpack = require("../../../../");
+const { VirtualUrlPlugin } = webpack.experiments.schemes;
+
+const watchDir = path.join(__dirname, "./routes");
+
+/** @type {import('webpack').Configuration} */
+const config = {
+	plugins: [
+		new VirtualUrlPlugin({
+			routes(loaderContext) {
+				const files = fs.readdirSync(watchDir);
+				return `
+					export const routes = {
+						${files.map(key => `${key.split(".")[0]}: () => import('./routes/${key}')`).join(",\n")}
+					}        
+				`;
+			},
+			app: "export const app = 'app'",
+			config: {
+				type: ".json",
+				source() {
+					return `{"name": "virtual-url-plugin"}`;
+				}
+			},
+			style: {
+				type: ".css",
+				source() {
+					return `body{background-color: powderblue;}`;
+				}
+			}
+		})
+	],
+	experiments: {
+		css: true
+	}
+};
+
+module.exports = config;
```

Comment:
Let's add a test case where we use a custom loader (babel for example for test) for generated code - we have https://webpack.js.org/configuration/module/#rulescheme and test, so I think it will work, just let;s ad test case

Reply:
Also let's add a test case for watching - something simple

Reply:
Test for cache invalidation will be great here too (we can use hmr test cases here), just read code from external file and put it directly in code

Reply:
> Let's add a test case where we use a custom loader (babel for example for test) for generated code - we have https://webpack.js.org/configuration/module/#rulescheme and test, so I think it will work, just let;s ad test case

Sorry for replying late. Thanks for your suggestions. I've now added use cases for a custom loader and asset/source.
https://github.com/webpack/webpack/pull/19508/files#diff-c3b387c10fec221ad0d92c4f1b3ea268e27065ab4dbc330a69ddfcc5d27d722bR27

Reply:
> Test for cache invalidation will be great here too (we can use hmr test cases here), just read code from external file and put it directly in code

And a test case for watching, with content read from an external file.
https://github.com/webpack/webpack/pull/19508/files#diff-6060d03f2a91f56b93302860b771391b8bc6511144fab0a7cd80e326c9fbb1dcR14